File: scala/difference-of-squares/src/main/scala/Squares.scala
class Squares {
  def squareOfSums(n: Int): Int = {
    val sum = (1 to n).sum
    sqr(sum)
  }

  def sumOfSquares(n: Int): Int = {
    val squares = (1 to n) map sqr
    squares.sum
  }

  def difference(n: Int): Int =
    squareOfSums(n) - sumOfSquares(n)

  private def sqr(n: Int) = n * n
}

object Squares {
  def apply() = new Squares
}

File: haskell/difference-of-squares/Squares.hs
module Squares (sumOfSquares, squareOfSums, difference) where

sumOfSquares :: (Integral a) => a -> a
sumOfSquares n = sum $ map (^2) [1..n]

squareOfSums :: (Integral a) => a -> a
squareOfSums n = (^2) $ sum [1..n]

difference :: (Integral a) => a -> a
difference n = squareOfSums n - sumOfSquares n
